### ISSUE TITLE
Validate compatibility with AWS Directory Service

### ISSUE THREAD

**Issue by aareet** [hashicorp/terraform-provider-ad#40]
### Description

- [ ] Check if the provider is compatible with AWS Directory Service (https://aws.amazon.com/directoryservice/)
- [ ] If it is, create an example to demonstrate to users how to quickly set up AD with AWS and use this provider to manage it
- [ ] If it is, add acceptance test coverage to cover this usage

### References

* https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/directory_service_directory

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by koikonom** [comment]
AWS provides two Directory services. Unfortunately the provider will not work at all with the "Basic Directory" service, because it's essentially a managed Samba instance. The other service they provide is the "Managed AD" service which is a managed Active Directory cluster. The problem with this one is that it does not come with WinRM enabled, nor does it allow you to enable it.

**Comment by tomwerneruk** [comment]
I am currently evaluating this provider, like you say @koikonom Managed AD is effectively a blackbox. My solution so far is to use a companion EC2 administration host (which is AWS recommended for operational use anyway).

**Comment by koikonom** [comment]
Hi @tomwerneruk, This is what I was thinking to try next. I just don't know if the powershell commands the provier runs will work "as is" or if they require any changes (Such as add a -ComputerName parameter or something).

**Comment by bart-parka** [comment]
I too am trying to get this to work with AWS Managed AD and a separate admin host. I believe I have managed to connect via Kerberos authentication but am now facing the following error:

`Error: command Get-ADUser exited with a non-zero exit code 1, stderr: #< CLIXML
<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><S S="Error">Get-ADUser : Unable to contact the server. This may be because this server does not exist, it is currently down, or it _x000D__x000A_</S><S S="Error">does not have the Active Directory Web Services running._x000D__x000A_</S><S S="Error">At line:1 char:1_x000D__x000A_</S><S S="Error">+ Get-ADUser -identity "S-1-5-21-3057401248-2679340083-1116673353-1286" ..._x000D__x000A_</S><S S="Error">+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~_x000D__x000A_</S><S S="Error">    + CategoryInfo          : ResourceUnavailable: (S-1-5-21-305740...1116673353-1286:ADUser) [Get-ADUser], ADServerDo _x000D__x000A_</S><S S="Error">   wnException_x000D__x000A_</S><S S="Error">    + FullyQualifiedErrorId : ActiveDirectoryServer:0,Microsoft.ActiveDirectory.Management.Commands.GetADUser_x000D__x000A_</S><S S="Error"> _x000D__x000A_</S></Objs>`

I have logged in as my user and opened up a Powershell session to confirm I have access to AD tools so I presume the problem is something to do with opening up a remote PS session and that not having necessary permissions.

**Comment by Type1J** [comment]
I would like to run this provider from my laptop, which is not in the AWS private network. The typical way to get into a VPC network isn't by VPN; it's by `ssh` or RDP through a bastion host. from the bastion host, which is probably joined to the domain, the provider could get access to the AD controller, but not with WinRM.

Are there plans to make this work? I would imagine that whatever solution is used for AWS Managed AD would work with other cloud providers.

**Comment by SkUrRiEr** [comment]
For future reference (mostly mine) the canonical way to manage AWS Managed Microsoft AD services is using RSAT: https://docs.aws.amazon.com/directoryservice/latest/admin-guide/microsoftadbasestep3.html (see the last step which essentially combines other `Install-WindowsFeature` commands elsewhere in the documentation) which is also the canonical way to manage Samba AD-DC servers from Windows. https://wiki.samba.org/index.php/Installing_RSAT (The Powershell command in the AWS documentation and the `dism` command in the Samba documentation are functionally equivalent)

So if we can update this provider to be able to be configured to talk the appropriate RSAT protocols, we should be able to manage AWS AD servers and Samba AD-DC servers assuming we can get single-hop access.

**Comment by Type1J** [comment]
It would probably be a safe bet that if you're modifying the AD, then you have AD credentials. If there needs to be a hop, then a bastion server that's running ssh (which could be locked to only talk to the internal network and a few trusted IPs) that is joined to the domain and could start an ssh tunnel using AD credentials could be assumed to exist in some hop scenarios. The credentials may be separate from the AD credentials as well, if a general solution is desired.

Basically, what I just said is like many database GUI clients that have an option to do ssh tunneling before talking to the database instead of directly opening a connection to the database.

**Comment by Type1J** [comment]
@SkUrRiEr I agree, the provider should speak a protocol that the AWS Managed AD understands. It may be that the RSAT actually uses WinRM.

**Comment by SkUrRiEr** [comment]
@Type1J, my understanding is that RSAT (which I think refers to the Tools on Windows, not the actual protocol) pre-dates WinRM. I'd expect that the Powershell management commands (e.g. `Get-ADUser`) communicate with the AD server on Windows using the same protocol as the RSAT tools.

**Comment by Type1J** [comment]
@SkUrRiEr I hope so. Is the protocol documented?

**Comment by koikonom** [comment]
Hello, 

Since WinRM connectivity on an AWS managed DC is not possible we would have to somehow work through an intermediate host. Since this has been requested a few times now I've opened a new issue to discuss about it.

I would love to hear your thoughts on https://github.com/hashicorp/terraform-provider-ad/issues/99 .

**Comment by koikonom** [comment]
Quick update on this issue. Starting with version v0.4.3 it is now possible to use most of the provider's functionality with the AWS managed AD service by leveraging the "double hop" authentication method. See [this section](https://registry.terraform.io/providers/hashicorp/ad/latest/docs#double-hop-authentication) in the provider docs that explains how double hop authentication is expected to work.

Pending work:
- The GPO related functionality will not work. This is still work in progress.

**Comment by shubhambjadhavar** [comment]
Can you provide an documentation with steps to configure this provider to manage AWS Managed Active Directory Remotely?
